src/background/stats.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * Ghostery Browser Extension
+ * https://www.ghostery.com/
+ *
+ * Copyright 2017-present Ghostery GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0
+ */
+import AnonymousCommunication from '@whotracksme/webextension-packages/packages/anonymous-communication';
+import Storage from '@whotracksme/webextension-packages/packages/anonymous-communication/storage';
+import {
+  getTimeAsYYYYMMDDHH,
+  getTimeAsYYYYMMDD,
+  getTrustedUtcTime,
+} from '@whotracksme/webextension-packages/packages/anonymous-communication/timestamps';
+
+const HPN_URL = 'https://collector-hpn.ghostery.net';
+const CONFIG_URL = 'https://api.ghostery.net/api/v1/config';
+const HPN_PUBLIC_KEYS_INDEXED_DB = 'hpn-public-keys';
+const STORAGE_KEY = 'stats';
+
+async function send(protocol) {
+  const now = getTrustedUtcTime();
+  const ts = getTimeAsYYYYMMDD(now);
+  const t = getTimeAsYYYYMMDDHH(now);
+
+  const data = await chrome.storage.local.get([STORAGE_KEY]);
+
+  if (t <= data[STORAGE_KEY].sentAt) {
```

Comment:
Would recommend to persist the UTC timestamp. Otherwise, you will compare strings here. Also, it introduces an unexpected, but nasty side-effect that all clients will send their alive signal at the start of the new hour. That leads to traffic spikes on the server.

Reply:
For now, the `sentAt` keeps the time rounded to the hour. The time when that condition will be checked is not at the start of the hour, but at a random time when the background process will start (or in the next hour from that random time). Because of that, there is no risk to have spikes in traffic on the server.

The only thing here is that the condition compares strings with iso date rather than numbers (as I thought, that function returns numbers). I will use UTC, but only because of that :)

Reply:
so when run at 11:59 and 12:01 it will send twice? we should rather check if there was 60 minutes in between the sends

Reply:
It might not look like a problem if you have, let's say thousands of client. But if you have users in the range of millions like in the Ghostery Desktop extension, the effects when messages are biased towards start-of-hour or start-of-day can be like a DDoS in the worst case, and in less extreme cases, it will at least make auto-scaling work very poorly (traffic hits systems that are scaled down, then race to scale up when the traffic is already going down again).